doc/designs/quic-design/connection-state-machine.md

```diff
@@ -0,0 +1,903 @@
+    - Are we currently doing proactive version negotiation?
+    - Have Successfully Received Any Encrypted Packet?
+    - Handshake Completed?
+    - Handshake Confirmed?
+
+  - The following timers are modelled:
+    - PTO Timeout
+    - Terminating Timeout
+    - Idle Timeout
+
+Implementation Plan
+-------------------
+
+- Phase 1: “Steady state only” model which jumps to the `ACTIVE.OPEN`
+  state with a hardcoded key.
+
+  Test plan: Currently uncertain, to be determined.
+
+- Phase 2: “Dummy handshake” model which uses a one-byte protocol
+  as the handshake layer as a standin for TLS 1.3. e.g. a 0x01 byte “represents”
+  a ClientHello, a 0x02 byte “represents” a ServerHello. Keys are fixed.
+
+  Test plan: If feasible, an existing QUIC implementation will be modified to
+  use this protocol and E2E testing will be performed against it. (This
+  can probably be done quickly but an alternate plan may be required if
+  the effort needed turns out be excessive.)
+
+- Phase 3: Final model with TLS 1.3 handshake layer fully plumbed in.
+
+  Test plan: Testing against real world implementations.
+
+Handling of Network I/O
+-----------------------
+
+Our API objectives make handling network I/O tricky.
+Broadly speaking:
+
+ - We want to support both blocking and non-blocking semantics
+   for application use of the libssl APIs.
+
+ - In the case of non-blocking applications, it must be possible
+   for an application to do its own polling and make its own event
+   loop.
+
+Moreover, traditional use of the libssl API allows an application to pass an
+arbitrary BIO to an SSL object; not only that, separate BIOs can be passed for
+the read and write directions. The nature of this BIO can be arbitrary; it could
+be a socket, or a memory buffer.
+
+Implementation of QUIC requires handling of timer events as well as the
+circumstances where a network socket becomes readable or writable. In many cases
+we need to handle these events simultaneously (e.g. wait until a socket becomes
+readable, or a timeout expires, whichever comes first).
+
+### Use of non-blocking I/O
+
+These requirements make it more or less a requirement that we use non-blocking
+network I/O; we need to be able to (at a minimum) have timeouts on recv() calls,
+and make best effort (non blocking) send() and recv() calls.
+
+The only sensible way to do this portably is to configure the socket into
+non-blocking mode. We could try to do a select() before calling send() or recv()
+to get a gurantee that the call will not block, but this will probably run into
+issues with buggy OSes which generate spurious socket readiness events. In any
+case, relying on this to work reliably is not advisable.
+
+Timeouts could be handled via setsockopt() socket timeout options, but this
+depends on OS support and adds another syscall to every network I/O operation.
+It also has obvious thread safety concerns if we want to move to concurrent use
+of a single socket at some later date.
```

Comment:
We have to get the API right now. There is no point trying to avoid e.g. using non-blocking mode if we are worried it will cause some application issue if we just end up having to do it later. I'd rather do it now when there are no existing users of our QUIC API. It will cause less pain.

Reply:
I agree with Hugo here. IMO trying to accompany blocking network sockets with QUIC is road bound to failure. And I do not even see a need for it. We can IMO still support blocking SSL_read()/SSL_write() calls with non-blocking network sockets, that should be possible to implement. Requiring the application to use non-blocking sockets when using QUIC should not require big changes in the application if we still allow for blocking SSL_read()/SSL_write().

Reply:
Our calls to the BIO should never block. The socket behind the BIO could be blocking, but the BIO needs to guarantee that we don't hang in the call.

Reply:
@t8m Right, my CSM PR as-is already implements blocking on top of non-blocking I/O. The application can get the blocking semantics it wants.

The only time an application would want to do its own I/O on a network socket also being used for QUIC is if it is muxing the socket with RTP/RTCP/STUN/etc. or similar using the QUIC fixed bit or something. But even then a) that is not really possible on the RX side without coordination with our QUIC stack because otherwise the application will receive datagrams intended for QUIC, so it would need to be a feature we add support for, b) there is no point doing it on the TX side if you can't also support RX, so no applications are going to be able to do this yet anyway, c) I would be very surprised if any applications working with RTP or similar are using blocking I/O anyway.

We've done a lot of work on adding `BIO_sendmmsg` and `BIO_recvmmsg` to the BIO APIs and implementing support across a lot of OSes, taking particular care to ensure these functions can be used concurrently in the future and do not mutate the state of the BIO. If we turn around and say that we're going to use the socket in blocking mode, we're forced to use socket options to set timeouts and all of that flies out the window.
In fact it would make our previous work on `BIO_sendmmsg` and `BIO_recvmmsg` largely pointless.

Reply:
You are missing the concept here - in that we explicitly noted wanting to be able to handle **both** approaches so we aren't forcing a particular model on applications. We want to support blocking and non-blocking and those applications that handle the network layer themselves and we handle the processing. It is one of the many ways this approach is different to other QUIC approaches in that we are not ramming a specific singular approach on the application.

Reply:
@t-j-h this design allows for all of this, except it requires the application to cope with the fact that the socket must be non-blocking. For applications that handle the network layer themselves it is applied appropriately - i.e., these applications must be able to not block on the BIO_recvmmsg/sendmmsg() (or BIO_read/write()) on the BIO that they provide to us.

Frankly, I cannot imagine any advanced application using blocking sockets. And I cannot imagine any simple application that currently uses blocking sockets that would be broken by making the blocking sockets non-blocking, if we still provide the blocking semantics on the SSL_read/write() calls.

Reply:
One use case that's probably not mentioned yet is that some people might be interested is using io_uring or ioring. io_uring recently got a zerocopy option, where it stays in user memory until the NIC says the packet has been transmitted. As far as I know, supporting minimal amount of copying was one of our design goals. But supporting io_uring like that probably requires a redesign of our current BIOs, which is probably something that has to wait until after 3.2.

Reply:
It seems the whole blocking/non-blocking thing needs some further discussion at OTC